atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/postgres/PostgresGetCandidateCellsForSweeping.java

```diff
@@ -0,0 +1,326 @@
+                    } else {
+                        computeNextStartPosition();
+                    }
+                    cellTsPairsExaminedTotal += cellTsPairsExaminedCurrentBatch;
+                    return results;
+                }
+            }
+        }
+
+        private void computeNextStartPosition() {
+            long lastTs = currentCellTimestamps.get(currentCellTimestamps.size() - 1);
+            // This can never happen because we request 'WHERE ts < ?', where '?' is some 'long' value.
+            // So if the timestamp is strictly less than another 'long', it can not be Long.MAX_VALUE.
+            // But we check anyway for general paranoia reasons.
+            Preconditions.checkState(lastTs != Long.MAX_VALUE);
+            firstCellStartTimestampInclusive = lastTs + 1;
+        }
+
+        private Optional<CandidateCellForSweeping> getCurrentCandidate() {
+            if (currentCellTimestamps.isEmpty()) {
+                return Optional.empty();
+            } else {
+                return Optional.of(ImmutableCandidateCellForSweeping.builder()
+                        .cell(Cell.create(currentRowName, currentColName))
+                        .sortedTimestamps(getSortedTimestamps())
+                        .isLatestValueEmpty(currentIsLatestValueEmpty)
+                        .numCellsTsPairsExamined(cellTsPairsExaminedTotal + cellTsPairsExaminedCurrentBatch)
+                        .build());
+            }
+        }
+
+        private long[] getSortedTimestamps() {
+            if (isCandidate()) {
+                long[] sortedTimestamps = currentCellTimestamps.toArray();
+                Arrays.sort(sortedTimestamps);
+                return sortedTimestamps;
+            } else {
+                return emptyLongArray;
+            }
+        }
+
+        private boolean isCandidate() {
+            return currentCellTimestamps.size() > 1
+                    || currentCellTimestamps.get(currentCellTimestamps.size() - 1)
+                            >= request.minUncommittedStartTimestamp()
+                    || currentCellTimestamps.contains(Value.INVALID_VALUE_TIMESTAMP)
+                    || currentIsLatestValueEmpty;
+        }
+
+        private ClosableIterator<AgnosticLightResultRow> selectNextPage(ConnectionSupplier conns) {
+            FullQuery query = getQuery();
+            AgnosticLightResultSet rs = conns.get().selectLightResultSetUnregisteredQuery(
+                    query.getQuery(), query.getArgs());
+            return ClosableIterators.wrap(rs.iterator(), rs);
+        }
+
+        private FullQuery getQuery() {
+            RangeBoundPredicates bounds = RangeBoundPredicates.builder(false)
+                    .startCellTsInclusive(currentRowName, currentColName, firstCellStartTimestampInclusive)
+                    .build();
+            boolean ignoreSentinels = areSentinelsIgnored();
+            String query = "/* GET_CANDIDATE_CELLS_FOR_SWEEPING(" + tableName + ") */"
```

Comment:
nice job here, this is actually fairly readable for its complexity.

My only complaint is that cell_ts_number isn't the greatest variable name for the subquery's max row number; (it made me think maybe this variable referred to the numeric values of a cell's logical timestamp); `cell_ts_count` would be better name